Add run property to latest task

helfi92 opened the pull request:
This will help remove additional requests from the workers page:

https://github.com/taskcluster/taskcluster-tools/blob/e8b6d45f10e7520f717b7a9f5db87d550c74d15e/src/views/Workers/WorkerManager.jsx#L123-L128.

eliperelman commented on src/graphql/Workers.graphql:
```diff
@@ -1,6 +1,7 @@
 type LatestTask {
   taskId: ID!
   runId: Int!
+  status: TaskRun!
```
Shouldn't `status` be a `TaskStatus`? If it needs to be a `TaskRun`, we should name the property accordingly.

eliperelman commented on src/loaders/workers.js:
```diff
@@ -15,6 +15,24 @@ export default ({ queue }) => {
     async ({ provisionerId, workerType, options, filter }) => {
       const raw = await queue.listWorkers(provisionerId, workerType, options);
+
+      raw.workers = await Promise.all(
```
While this technically works, it bypasses a lot of the logic we have surrounding loading statuses.

What you want to do instead is override the resolver on a `LatestTask` to return a `TaskStatus` at that property:

```js
// in https://github.com/taskcluster/taskcluster-web-server/blob/master/src/resolvers/Workers.js
LatestTask: {
  status(parent, args, { loaders }) {
    return loaders.status.load(parent.taskId);
  }
}
```

If instead of getting the status you want the actual run, rename to `run`, and pull the run out (since I don't think there is a TC API call for just getting a single run):

```js
LatestTask: {
  run(parent, args, { loaders }) {
    return loaders.status.load(parent.taskId).runs
      .find(run => run.runId === parent.runId);
  }
}
```

Something like that.

helfi92 commented:
@eliperelman I refactored to have the logic in the resolver. I also added `taskGroupId`. Is there a way to remove the duplicate call to `loaders.status.load` without having to merge `taskGroupId` inside the `run` property?

helfi92 commented:
@eliperelman Removed the compact worker but `listWorkers` returns a list of compact workers so I had to change the loader to loop over that list and load the full worker instead. I'm not sure if that logic should be in the loader or resolvers. I put it in the loader for now. Let me know what you recommend.

eliperelman commented:
It won't ever paginate since this API doesn't return a `continuationToken`. The `ConnectionLoader` is only for paginated connections.

eliperelman commented:
Wait, I misread that.

helfi92 commented:
As discussed on IRC, I went back to the original plan of adding a `run` property.

helfi92 commented:
Ah, I think there's something still missing in my implementation. Looking into it.

helfi92 commented:
Done.

eliperelman commented on src/resolvers/Workers.js:
```diff
@@ -1,4 +1,9 @@
 export default {
+  LatestTask: {
+    async run(parent, args, { loaders }) {
+      return (await loaders.status.load(parent.taskId)).runs[parent.runId];
```
Can you break this up into 2 lines?

```js
const status = await loaders.status.load(parent.taskId);

return status.runs[parent.runId];
```
